homeassistant/components/deluge/__init__.py

```diff
@@ -1 +1,356 @@
+        self._deluge_data = None
+        self.unsub_timer = None
+
+    @property
+    def api(self):
+        """Return the _deluge_data object."""
+        return self._deluge_data
+
+    async def async_setup(self):
+        """Set up the Deluge client."""
+
+        try:
+            deluge_api = await get_api(self.hass, self.config_entry.data)
+        except CannotConnect:
+            raise ConfigEntryNotReady
+        except (UserNameError, PasswordError, UnknownError):
+            return False
+
+        self._deluge_data = DelugeData(self.hass, self.config_entry, deluge_api)
+
+        await self.hass.async_add_executor_job(self._deluge_data.init_torrent_list)
+        await self.hass.async_add_executor_job(self._deluge_data.update)
+        self.add_options()
+        self.set_scan_interval(self.config_entry.options[CONF_SCAN_INTERVAL])
+
+        for component in PLATFORMS:
+            self.hass.async_create_task(
+                self.hass.config_entries.async_forward_entry_setup(
+                    self.config_entry, component
+                )
+            )
+
+        def add_torrent(service):
+            """Add new torrent to download."""
+            torrent = service.data[ATTR_TORRENT]
+            deluge_client = None
+            for entry in self.hass.config_entries.async_entries(DOMAIN):
+                if entry.data[CONF_HOST] == service.data[CONF_HOST]:
+                    deluge_client = self.hass.data[DOMAIN][entry.entry_id]
+                    break
+            if deluge_client is None:
+                _LOGGER.error("Deluge host is not found")
+                return
+            if torrent.startswith(("http", "ftp:")):
+                deluge_api.core.add_torrent_url(torrent, {})
+            elif torrent.startswith(("magnet:")):
+                deluge_api.core.add_torrent_magnet(torrent, {})
+            else:
+                _LOGGER.warning("Could not add torrent: unsupported type")
+
+        self.hass.services.async_register(
+            DOMAIN, SERVICE_ADD_TORRENT, add_torrent, schema=SERVICE_ADD_TORRENT_SCHEMA
+        )
+
+        self.config_entry.add_update_listener(self.async_options_updated)
+
+        return True
+
+    def add_options(self):
+        """Add options for entry."""
+        if not self.config_entry.options:
+            scan_interval = self.config_entry.data.pop(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            )
+            options = {CONF_SCAN_INTERVAL: scan_interval}
+
+            self.hass.config_entries.async_update_entry(
+                self.config_entry, options=options
+            )
+
+    def set_scan_interval(self, scan_interval):
+        """Update scan interval."""
+
+        def refresh(event_time):
+            """Get the latest data from Transmission."""
+            self._deluge_data.update()
+
+        if self.unsub_timer is not None:
+            self.unsub_timer()
+        self.unsub_timer = async_track_time_interval(
```

Comment:
async in sync.

Reply:
If this is supposed to be an async method, name it so and annotate it with `@callback` (imported from `homeassistant.core`)

Reply:
The issue here is that deluge api is not async and we have a `refresh` method under `set_scan_interval` that calls for update.
So would you recommend using `track_time_interval` instead or making `set_scan_interval` as async?

Reply:
if you can, try to do as much in async land, only using executor to go into sync land when necessary.

Reply:
@MartinHjelmare  just highlighted that `add_options` method is called from async context. The same is done for the `set_scan_interval` method.

Reply:
To make it easier for readers to follow, I recommend renaming these methods and prefix them with `async_`. That's enough I think. If we might schedule them with a listener we need to decorate them with `@callback` too.